### PR TITLE
WebAssembly module

### PR DESCRIPTION
**Pull request by adzialocha**
- Message encoding now has it's own method `encode_create_message()` which returns `message_encoded` as a string
- `sign_entry()` (previously `sign_encode()`) now takes the encoded message and extra argument log_id  returns `entry_encoded` and `entry_hash` 
- `decode_entry()` now returns an `entry` as lovely json object
- Nicer error handling :+1: 

Closes: #39, #48 and #51

### CODE REVIEW

**Comment by sandreae** [comment]
This looks great! Just getting my head round what's changed..

- message encoding now has it's own method `encode_create_message()` which returns `message_encoded` as a string
- `sign_entry()` (previously `sign_encode()`) now takes the encoded message and extra argument log_id  returns `entry_encoded` and `entry_hash` 
- `decode_entry()` now returns an `entry` as lovely json object
- nicer error handling :+1: 

What was the reason for splitting the message encoding? I like that it's more explicit about what is going on, and separation of tasks. Just wondered if there were other reasons too.

**Comment by adzialocha** [comment]
Thanks for summarizing, I've added it now in the description of this PR :+1: 

> What was the reason for splitting the message encoding? I like that it's more explicit about what is going on, and separation of tasks. Just wondered if there were other reasons too.

There will be different sorts of messages (Create, Update, Delete), so I'm thinking about introducing also an `encode_update_message` and `encode_delete_message` which results you can then pass over to `sign_entry`.

**Comment by sandreae** [comment]
> There will be different sorts of messages (Create, Update, Delete), so I'm thinking about introducing also an encode_update_message and encode_delete_message which results you can then pass over to sign_entry.

Ok, thought there might be something deeper ;-p that makes sense!

**Comment by adzialocha** [comment]
> Ok, thought there might be something deeper ;-p that makes sense!

Haha no! Also, please feel free to change the API, this is really just a draft yet.

**Comment by sandreae** [comment]
Maybe we can set aside some time to chat together about it at one of the next meetups?

File: p2panda-rs/src/wasm.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    pub fn new() -> Self {
+        Self(PandaMessageFields::new())
+    }
+
+    pub fn add(&mut self, name: String, value: JsValue) -> Result<(), JsValue> {
+        // @TODO: Add more types
+        let field = match value.as_string() {
+            Some(text) => Ok(MessageValue::Text(text)),
+            None => Err(js_sys::Error::new(&format!("Invalid value type"))),
+        }?;
```
Comment (cafca) [inline p2panda-rs/src/wasm.rs:44]: Frustrating not to have  schema definitions for this.

File: p2panda-rs/src/wasm.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        jserr!(self.0.add(&name, field));
+
+        Ok(())
+    }
+}
+
+#[wasm_bindgen(js_name = encodeCreateMessage)]
+pub fn encode_create_message(schema: String, fields: MessageFields) -> Result<String, JsValue> {
```
Comment (cafca) [inline p2panda-rs/src/wasm.rs:53]: I like the functional style here.

File: p2panda-rs/src/wasm.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        seq_num.as_ref(),
+    ));
+
+    // Finally sign and encode entry
+    let entry_signed = jserr!(EntrySigned::try_from((&entry, &key_pair)));
+
+    // Serialize result to JSON
+    let result = jserr!(wasm_bindgen::JsValue::from_serde(&SignEntryResult {
+        entry_encoded: entry_signed.as_str().into(),
+        entry_hash: entry_signed.hash().as_hex().into(),
```
Comment (cafca) [inline p2panda-rs/src/wasm.rs:113]: Should we use either `as_str` or `as_hex` throughout the library for consistency?

File: p2panda-rs/src/wasm.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct SignEntryResult {
+    pub entry_encoded: String,
+    pub entry_hash: String,
+}
+
+#[wasm_bindgen(js_name = signEntry)]
+pub fn sign_entry(
+    key_pair: KeyPair,
```
Comment (sandreae) [inline p2panda-rs/src/wasm.rs:69]: I think this is causing some issues on the JS end of things. After passing a KeyPair instance into signEntry() it no longer exists on the JS end. I guess this is a rust ownership thing. I tried:

`key_pair: &KeyPair,`

And that works, but I don't know if it is strange Rust....

Right now I am just passing in a copy of the keyPair which also works.

**Comment by cafca** [comment]
LOL

Sam Andreae ***@***.***> schrieb am Mi. 7. Apr. 2021 um 17:46:

File: p2panda-rs/src/wasm.rs (hunk 3 above)
Comment (adzialocha) [reply p2panda-rs/src/wasm.rs:113]: Yes, but I guess thats part of another issue?

File: p2panda-rs/src/wasm.rs (hunk 4 above)
Comment (adzialocha) [reply p2panda-rs/src/wasm.rs:69]: > And that works, but I don't know if it is strange Rust....

Looks like a good solution! Passing over instances via reference is even better here. Do you want to make a commit?

File: p2panda-rs/src/wasm.rs (hunk 3 above)
Comment (cafca) [reply p2panda-rs/src/wasm.rs:113]: That's what I was thinking

Comment (cafca) [reply p2panda-rs/src/wasm.rs:131]: https://github.com/p2panda/p2panda/issues/50

**Review by cafca — APPROVED** [review]

**Pull request merged**